File: flask_app/app/controllers/user_tier_metrics_controller.py
from datetime import datetime
from flask import request
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.services.user_tier_metrics_service import UserTierMetricsService
from app.utils.response import success_response, error_response


class UserTierMetricsController:

    @staticmethod
    @jwt_required()
    def get_my_metrics(month_year):
        user_id = get_jwt_identity()
        metrics = UserTierMetricsService.get_for_user_month(user_id, month_year)
        if not metrics:
            return error_response("Metrics not found", 404)
        return success_response("Metrics retrieved", metrics.to_dict())

    @staticmethod
    @jwt_required()
    def create_or_update(month_year):
        user_id = get_jwt_identity()
        data = request.get_json()
        required_fields = ["order_count", "total_spent"]

        # Validate required fields
        missing = [f for f in required_fields if f not in data]
        if missing:
            return error_response(f"Missing fields: {', '.join(missing)}", 400)

        # Validate month format
        try:
            datetime.strptime(month_year, "%Y-%m")
        except ValueError:
            return error_response("Invalid month format. Use YYYY-MM.", 400)

        data.update({"user_id": user_id, "month_year": month_year})
        utm = UserTierMetricsService.create_or_update(data)
        return success_response("Metrics updated", utm.to_dict())

    @staticmethod
    @jwt_required()
    def get_user_metrics():
        """Get metrics for all users (admin/reporting purpose)."""
        all_metrics = UserTierMetricsService.get_all_metrics()
        return success_response(
            "All metrics retrieved", [m.to_dict() for m in all_metrics]
        )

    @staticmethod
    @jwt_required()
    def get_top_spenders():
        """Get top N spenders for a given month. Defaults to current month."""
        month_year = request.args.get("month_year")
        top_n = int(request.args.get("top_n", 10))
        if not month_year:
            month_year = datetime.utcnow().strftime("%Y-%m")

        top_spenders = UserTierMetricsService.get_top_spenders(month_year, top_n)
        return success_response(
            "Top spenders retrieved", [m.to_dict() for m in top_spenders]
        )

File: springboot_api/src/main/java/com/example/membership_api/controller/UserTierMetricsController.java
package com.example.membership_api.controller;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.membership_api.model.UserTierMetrics;
import com.example.membership_api.service.UserTierMetricsService;
import com.example.membership_api.utils.JwtUtil;
import com.example.membership_api.utils.ResponseUtil;

import jakarta.servlet.http.HttpServletRequest;

@RestController
@RequestMapping("/user-metrics")
public class UserTierMetricsController {

    @Autowired
    private UserTierMetricsService service;

    @GetMapping("/{monthYear}")
    public ResponseEntity<?> getMyMetrics(@PathVariable String monthYear, HttpServletRequest request) {
        String userId = JwtUtil.extractUserIdFromRequest(request);
        UserTierMetrics metrics = service.getForUserMonth(userId, monthYear);
        if (metrics == null) {
            return ResponseUtil.error("Metrics not found", 404);
        }
        return ResponseUtil.success("Metrics retrieved", metrics.toMap());
    }

    @PostMapping("/{monthYear}")
    public ResponseEntity<?> createOrUpdate(
        HttpServletRequest request,
            @RequestBody Map<String, Object> payload
    ) {
        String userId = JwtUtil.extractUserIdFromRequest(request);
        if (!payload.containsKey("order_count") || !payload.containsKey("total_spent")) {
            return ResponseUtil.error("Missing fields: order_count, total_spent", 400);
        }

        int orderCount = (int) payload.get("order_count");
        double totalSpent = Double.parseDouble(payload.get("total_spent").toString());

        UserTierMetrics utm = service.createOrUpdate(userId, orderCount, totalSpent);
        return ResponseUtil.success("Metrics updated", utm.toMap());
    }

    @GetMapping
    public ResponseEntity<?> getAllMetrics() {
        List<UserTierMetrics> all = service.getAllMetrics();
        return ResponseUtil.success("All metrics retrieved", all.stream().map(UserTierMetrics::toMap).toList());
    }

    @GetMapping("/top_spenders")
    public ResponseEntity<?> getTopSpenders(
            @RequestParam(required = false) String monthYear,
            @RequestParam(defaultValue = "10") int topN
    ) {
        if (monthYear == null || monthYear.isEmpty()) {
            monthYear = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM"));
        }

        List<UserTierMetrics> top = service.getTopSpenders(monthYear, topN);
        return ResponseUtil.success("Top spenders retrieved", top.stream().map(UserTierMetrics::toMap).toList());
    }
}
